### PR TITLE
Upgrade Postgres Major Version

### PR DESCRIPTION
**Pull request by rc9000**
TL;DR: do not just merge this right now :)

I did some digging how to navigate an upgrade of the Postgres version, as 13 will be out of support next year. Unfortunately this is a bit of a support headache, since it needs manual converting of the Postgres files. I tried to include an easy one-command approach using the pgautoupgrade project: https://github.com/pgautoupgrade/docker-pgautoupgrade

How to use it is in the README_UPGRADE.md, which the PG 17 container also just dumps into the container log if there is a mismatch.

However most users probably start the container in the background, in Docker Desktop etc. and will still not immediately see it. So we should find a good time to create a new release in Netdisco proper, declare this the cutoff for netdisco-docker PG13, and then I'd hang out in the issues & IRC for a week or two, because people will probably need help :)

To try the instructions or just test Netdisco with PG17 in general, you can get the new container from my Dockerhub account - just use `rc9000/netdisco:latest-pg17` as a drop-in replacement for `netdisco:latest-postgresql`.

Actual EOL for PG13 is  November 13, 2025 so there is no rush.

### CODE REVIEW

**Comment by rc9000** [comment]
IRC log: after some discussions we decided that a separate container with pg17 would be nice, that the user can enable at will instead of being forced into it:


    14:01 < rc9000> one way i could see instead: we make an alternative netdisco-postgresql17 
                    container that includes the autoupgrade but is commented out in docker-compose, plus some notes that the user can
                    switch to it when they know they have a valid backup or don't care, or just want to start over with a new db?
    14:01 < rc9000> this would be nontrivial since we need to merge the whole pgautoupgrade into our entrypoint, but definitely doable.
    14:01 < rc9000> (the relevant bits are around here https://github.com/pgautoupgrade/docker-pgautoupgrade/blob/main/postgres-docker-entrypoint.sh#L567)
    15:01 < oliver> rc9000: yes I was thinking along these lines ... a straightforward thing the user could do (change a text config and restart) to opt in to the upgrade
    17:10 < rc9000> oliver: ok let's rescope the pr so there is a new container that can be enabled, and then we can slowly get people into switching over.
    17:11 < rc9000> that's also nice because there is no big bang
    17:14 < rc9000> should we just append the major version like so?
    17:14 < rc9000> netdisco:2.084002-postgresql17
    17:14 < rc9000> netdisco:latest-postgresql17
    17:14 < rc9000> ... time passes
    17:14 < rc9000> netdisco:2.084002-postgresql22
    17:14 < rc9000> netdisco:latest-postgresql22
    17:15 < rc9000> and we could also push the current one additionally as netdisco:latest-postgresql13 for symmetry
    17:52 < oliver> rc9000: yes would be OK but I'd like to check for anything like a loop over the suffix on the image name which would be caught out 
                    (like for i in do backend frontend postgresql...)

**Comment by rc9000** [comment]
I did some experiments with integration docker-pgautoupgrade directly into our postgres container: https://github.com/rc9000/netdisco-docker/tree/pg17

It works, but it has some major downsides:

1. Major: the biggest one is that the image must contain the old and the new postgres version. So once Postgres 13 is unsupported and has racked up some CVE that are no longer fixed, the image will be flagged in all the enterprise scanning tools. I don't know how many people this would affect, but it would certainly affect me because everybody here loves their grype and Artifactory reports and going after people with them.
2. Minor: image size is greatly increased
3. Major: pgautoupgrade can not guarantee to work and not corrput the db, and needs careful DBAing around it (disk space, backups etc.) to prevent that. The safest route is to let the user deal with that and be responsible, that's also the reason why the official postgres image doesn't have such features

So currently I think the original approach of this PR might be better.

File: netdisco-postgresql/netdisco-db-entrypoint.sh [hunk 1]
```diff
@@ -10,6 +10,11 @@ if [ ! -s "${PGDATA}/PG_VERSION" ]; then
     /usr/local/bin/docker-entrypoint.sh "$@"
 fi
 
+if ! grep -q "17" "${PGDATA}/PG_VERSION" ; then
```
[the remark this answers is not in the document]
Comment (rc9000) [reply netdisco-postgresql/netdisco-db-entrypoint.sh:13]: Hogwash, Clanker :) This is always major version in the format ^\d+$

**Comment by ollyg** [comment]
our collective bacon has been saved? [Postgres docker now uses version specific PGDATA location](https://github.com/docker-library/docs/blob/master/postgres/README.md#pgdata).

- meaning we can run two postgres instances side by side (separate data directories)
- we can construct a migration path for users. this is because Netdisco has a very predictable DB schema (if user has strayed from the schema then they take on all the management responsiblity).
- meaning we only need to copy the *data* not the schema, so no "major version migration".
- so have a docker compose profile or a mix-in compose file to temporarily run second postgres and migrate across

1. start both postgres (left and right)
2. bring both schema up to date (already have a solution for it)
4. pipe from pg_dump (left) to pg_restore or psql (right) (make sure to use the new version's pg_dump, it's guaranteed to work)

When the user updates their compose.yaml it'll be running the latest (18) postgres and an empty schema. They will have to run the mix in or profile to get the upgrade done. This means default start is always latest release.

**Comment by ollyg** [comment]
hopefully no longer needed
